cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KRaftMigrationUtils.java

```diff
@@ -31,41 +35,97 @@ public class KRaftMigrationUtils {
      * This method deletes the /controller znode from ZooKeeper to allow the brokers, which are now in ZooKeeper mode again,
      * to elect a new controller among them taking the KRaft controllers out of the picture.
      *
-     * @param reconciliation        Reconciliation information
-     * @param coTlsPemIdentity      Trust set and identity for TLS client authentication for connecting to ZooKeeper
-     * @param operationTimeoutMs    Timeout to be set on the ZooKeeper request configuration
-     * @param zkConnectionString    Connection string to the ZooKeeper ensemble to connect to
+     * @param reconciliation            Reconciliation information
+     * @param vertx                     Vert.x instance
+     * @param zooKeeperAdminProvider    ZooKeeper Admin client provider
+     * @param coTlsPemIdentity          Trust set and identity for TLS client authentication for connecting to ZooKeeper
+     * @param operationTimeoutMs        Timeout to be set on the ZooKeeper request configuration
+     * @param zkConnectionString        Connection string to the ZooKeeper ensemble to connect to
+     *
+     * @return Completes when the /controller znode deletion is done or any error
      */
-    public static void deleteZooKeeperControllerZnode(Reconciliation reconciliation, TlsPemIdentity coTlsPemIdentity, long operationTimeoutMs, String zkConnectionString) {
+    public static Future<Void> deleteZooKeeperControllerZnode(Reconciliation reconciliation, Vertx vertx, ZooKeeperAdminProvider zooKeeperAdminProvider, TlsPemIdentity coTlsPemIdentity, long operationTimeoutMs, String zkConnectionString) {
         // Setup truststore from PEM file in cluster CA secret
         File trustStoreFile = Util.createFileStore(KRaftMigrationUtils.class.getName(), PemTrustSet.CERT_SUFFIX, coTlsPemIdentity.pemTrustSet().trustedCertificatesPemBytes());
 
         // Setup keystore from PEM in cluster-operator secret
         File keyStoreFile = Util.createFileStore(KRaftMigrationUtils.class.getName(), PemAuthIdentity.PEM_SUFFIX, coTlsPemIdentity.pemAuthIdentity().pemKeyStore());
-        try (ZooKeeperAdmin admin = new DefaultZooKeeperAdminProvider().createZookeeperAdmin(
+
+        return connectToZooKeeper(reconciliation, vertx, zooKeeperAdminProvider, trustStoreFile, keyStoreFile, operationTimeoutMs, zkConnectionString)
+                .compose(zkAdmin -> {
+                    Promise<Void> znodeDeleted = Promise.promise();
+                    try {
+                        zkAdmin.delete("/controller", -1);
+                        LOGGER.infoCr(reconciliation, "Deleted the '/controller' znode as part of the KRaft migration rollback");
+                        znodeDeleted.complete();
+                    } catch (KeeperException | InterruptedException e)    {
```

Comment:
What if IllegalArgumentException is thrown?

Reply:
Which part can raise such exception here? I am catching what the `zkAdmin.delete` could throw.

Reply:
`delete(...)` has it in its Javadoc -> when the path does not exist. But I would maybe correct my self -> you should catch anything in here. Because if you don't, you will never complete the future and the reconciliation will likely get _stuck_, or?

Reply:
Didn't notice it on the Javadoc. But yeah you are right, it's better catching any exception.